packages/query-persist-client-core/src/createPersister.ts

```diff
@@ -0,0 +1,158 @@
+import { hashKey } from '@tanstack/query-core'
+import type {
+  QueryClient,
+  QueryFunctionContext,
+  QueryKey,
+  QueryState,
+} from '@tanstack/query-core'
+
+export interface PersistedQuery {
+  buster: string
+  queryHash: string
+  queryKey: QueryKey
+  state: QueryState
+}
+
+export interface AsyncStorage {
+  getItem: (key: string) => Promise<string | undefined | null>
+  setItem: (key: string, value: string) => Promise<void>
+  removeItem: (key: string) => Promise<void>
+}
+
+export interface StoragePersisterOptions<QC extends QueryClient> {
+  /**
+   * Query Client instance
+   */
+  queryClient: QC
+  /** The storage client used for setting and retrieving items from cache.
+   * For SSR pass in `undefined`.
+   */
+  storage: AsyncStorage | Storage | undefined | null
+  /**
+   * How to serialize the data to storage.
+   * @default `JSON.stringify`
+   */
+  serialize?: (client: PersistedQuery) => string
+  /**
+   * How to deserialize the data from storage.
+   * @default `JSON.parse`
+   */
+  deserialize?: (cachedString: string) => PersistedQuery
+  /**
+   * A unique string that can be used to forcefully invalidate existing caches,
+   * if they do not share the same buster string
+   */
+  buster?: string
+  /**
+   * The max-allowed age of the cache in milliseconds.
+   * If a persisted cache is found that is older than this
+   * time, it will be discarded
+   */
+  maxAge?: number
+}
+
+/**
+ * Warning: experimental feature.
+ * This utility function enables fine-grained query persistance.
+ * Simple add it as a `persister` parameter to `useQuery` or `defaultOptions` on `queryClient`.
+ * 
+ * ```
+ * useQuery({
+     queryKey: ['myKey'],
+     queryFn: fetcher,
+     persister: createPersister({
+       storage: localStorage,
+       queryClient,
+     }),
+   })
+   ```
+ */
+export function experimental_createPersister<T, QC extends QueryClient>({
+  queryClient,
```

Comment:
@TkDodo I think we would need a way to get access to `queryClient` somehow from `Query` context to not require it being passed by the user.
Otherwise how would you provide it as a default query option - chicken and egg problem.
```ts
new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 3600,
      persister: createPersister({
        storage: {
          getItem: (key: string) => get(key),
          setItem: (key: string, value: string) => set(key, value),
          removeItem: (key: string) => del(key),
        },
        queryClient, // ? but... how?
      }),
    }
  },
})
```

Reply:
true. it would work if it were a function I think:

```
const queryClient: QueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 3600,
      persister: () => createPersister({
        storage: {
          getItem: (key: string) => get(key),
          setItem: (key: string, value: string) => set(key, value),
          removeItem: (key: string) => del(key),
        },
        queryClient,
      }),
    }
  },
})
```

this would go well with my proposal of making persisters an optional function.

---

Oh, I just saw that we also expose a `setDefaultOptions` function from the `queryClient`:

https://github.com/TanStack/query/blob/9c721e548f500be924da8ef6da041a919f513140/packages/query-core/src/queryClient.ts#L617-L619

so you can really do:

```
const queryClient = new QueryClient()
queryClient.setDefaultOptions({
  queries: {
    staleTime: 3600,
    persister: createPersister({
      storage: {
        getItem: (key: string) => get(key),
        setItem: (key: string, value: string) => set(key, value),
        removeItem: (key: string) => del(key),
      },
      queryClient,
    }),
  }
})
```

and that should work without problems